### PR TITLE
set dirw to NULL to prevent freeing it multiple times

### LINKED ISSUE #1221: fs.watch crashes process if filename is a shared SharePoint folder
Initially reported as https://github.com/Microsoft/TypeScript/issues/13874
Turned out node crashes if fs.watch is called for a folder which is SharePoint shared folder.

At glance the problem seems to be in `fs-event.c`
- [here](https://github.com/libuv/libuv/blob/v1.x/src/win/fs-event.c#L189-L192) `uv_fs_event_start` sets `handle->dirw` to `pathw`
- then `ReadDIrectoryChangesW` [fails](https://github.com/libuv/libuv/blob/v1.x/src/win/fs-event.c#L258) and error handling logic [here](https://github.com/libuv/libuv/blob/v1.x/src/win/fs-event.c#L296) frees `pathw` making `handle-dirw` point to a released memory.
- as a final step `uv_fs_event_stop` [checks](https://github.com/libuv/libuv/blob/v1.x/src/win/fs-event.c#L338-L341) if `handle->dirw` is not null (which it true) and if yes - tries to free it which fails since it was already freed on the previous step

### PR DESCRIPTION
**Pull request by vladima**
fixes #1221

### CODE REVIEW

File: src/win/fs-event.c [hunk 1]
```diff
@@ -293,6 +293,11 @@ int uv_fs_event_start(uv_fs_event_t* handle,
     handle->short_filew = NULL;
   }
 
+  if (handle->dirw) {
+	  assert(is_path_dir);
+	  assert(handle->dirw == pathw);
+	  handle->dirw = NULL;
+  }
```
Comment (santigimeno) [inline src/win/fs-event.c]: style:
- 2 space indentation.
- add empty line after closing curly bracket

**Comment by vladima** [comment]
that is a good question. I was only able to repro this issue for SharePoint shared folders when control flow successfully passed check [here](https://github.com/vladima/libuv/blob/e112388d7bf0a5a1618af7257e53f8993bb36c8f/src/win/fs-event.c#L182) and failed only at [this line](https://github.com/vladima/libuv/blob/e112388d7bf0a5a1618af7257e53f8993bb36c8f/src/win/fs-event.c#L258). If test infrastructure for libuv allows to mimic this I'd be glad to add tests.

**Comment by vladima** [comment]
@santigimeno can you please tell what else needs to be done on this item? This is a pretty serious issue for TypeScript server since in cases of SharePoint shared folders mapped to drives there is no reasonable way from node to detect it.

**Comment by bnoordhuis** [comment]
Would this patch work?
```diff
diff --git a/src/win/fs-event.c b/src/win/fs-event.c
index 03e4adc..07f91e4 100644
--- a/src/win/fs-event.c
+++ b/src/win/fs-event.c
@@ -186,11 +186,10 @@ int uv_fs_event_start(uv_fs_event_t* handle,
 
   is_path_dir = (attr & FILE_ATTRIBUTE_DIRECTORY) ? 1 : 0;
 
   if (is_path_dir) {
      /* path is a directory, so that's the directory that we will watch. */
-    handle->dirw = pathw;
     dir_to_watch = pathw;
   } else {
     /*
      * path is a file.  So we split path into dir & file parts, and
      * watch the dir directory.
@@ -273,10 +272,11 @@ int uv_fs_event_start(uv_fs_event_t* handle,
     last_error = GetLastError();
     goto error;
   }
 
   handle->req_pending = 1;
+  handle->dirw = pathw;
   return 0;
 
 error:
   if (handle->path) {
     uv__free(handle->path);
```
If not, then your pull request is fine, but this patch probably looks more Obviously Correct to someone reading the source code.

**Comment by vladima** [comment]
in my PR I've tried to preserve exactly what existing code was doing. Proposed patch changes the behavior of existing code code in sense that now `handle-dirw` is only set when target path is directory and now it will be set all the time. At glance it should not make any differences but I'm not familiar with the codebase to state this for sure.

**Comment by vladima** [comment]
@bnoordhuis  to elaborate, I'm talking about the code like [this](https://github.com/libuv/libuv/blob/v1.x/src/win/fs-event.c#L401) that assumes if `dirw` is set then it is watching directory.
However I think if we make setting `dirw` in your patch conditional it will be the simplest correct solution:
```diff
diff --git a/src/win/fs-event.c b/src/win/fs-event.c
index 03e4adc..07f91e4 100644
--- a/src/win/fs-event.c
+++ b/src/win/fs-event.c
@@ -186,11 +186,10 @@ int uv_fs_event_start(uv_fs_event_t* handle,
 
   is_path_dir = (attr & FILE_ATTRIBUTE_DIRECTORY) ? 1 : 0;
 
   if (is_path_dir) {
      /* path is a directory, so that's the directory that we will watch. */
-    handle->dirw = pathw;
     dir_to_watch = pathw;
   } else {
     /*
      * path is a file.  So we split path into dir & file parts, and
      * watch the dir directory.
@@ -273,10 +272,11 @@ int uv_fs_event_start(uv_fs_event_t* handle,
     last_error = GetLastError();
     goto error;
   }
 
   handle->req_pending = 1;
+  if (is_path_dir) { 
+    handle->dirw = pathw;
+ }
   return 0;
 
 error:
   if (handle->path) {
     uv__free(handle->path);
```

**Comment by bnoordhuis** [comment]
Isn't `dirw == NULL` when `is_path_dir == 0`?  I would think the conditional isn't necessary.

**Comment by vladima** [comment]
yes, this is how it is works now. However if `dirw` will be set to `pathw` all the time this will no longer be true that is why I'm suggesting to make setting conditional so existing behavior is preserved

**Comment by vladima** [comment]
is there anything else that should be updated in this PR?

**Comment by santigimeno** [comment]
wrt a possible test, what about something like this so `ReadDirectoryChangesW` returns this error: 

> ReadDirectoryChangesW fails with ERROR_NOACCESS when the buffer is not aligned on a DWORD boundary.

```c
r = uv_fs_event_init(loop, &fs_event);
ASSERT(r == 0);
fs_event.buffer = (char*)malloc(1111);
r = uv_fs_event_start(&fs_event, fs_event_cb_dir_multi_file, "watch_dir", 0);
ASSERT(r == UV_EACCES);
r = uv_fs_event_stop(&fs_event);
ASSERT(r == 0);
```

**Comment by vladima** [comment]
sure, that should work! A question: does `(char*)malloc(1111)` is guaranteed to return non-DWORD aligned pointer or this is unspecified detail that can vary across implementations?

**Comment by vladima** [comment]
actually it feels that this MSDN statement is out-of-date.
> ReadDirectoryChangesW fails with ERROR_NOACCESS when the buffer is not aligned on a DWORD boundary.

At least I can see that with buffer `0x00691b08` (which does not seem to me as DWORD aligned) `ReadDirectoryChangesW` still does not fail.

**Comment by santigimeno** [comment]
@vladima you're right, it doesn't fail with unaligned buffer on my `Win10 VM`. The funny thing is that it fails with `ERROR_NOACCESS` in my box with a `malloc(1111)` which is indeed aligned. Strange.

**Comment by santigimeno** [comment]
@bnoordhuis are you fine with the PR as is?

**Comment by bnoordhuis** [comment]
I'd prefer the simpler no-conditionals solution.

> However if dirw will be set to pathw all the time this will no longer be true that is why I'm suggesting to make setting conditional so existing behavior is preserved

I appreciate your defensive coding approach but, if I understand you correctly, if and when we make a change like that, we'd have to review all the code paths in that function anyway.

**Comment by vladima** [comment]
@bnoordhuis I've checked the code and it looks that unconditional setting of `dirw` will work in all cases - when watching file [this bit](https://github.com/vladima/libuv/blob/5bab8f58946f06f438230a2879d174613e8c9dfa/src/win/fs-event.c#L216) will set `pathw` to `NULL` so at the end of the function `dirw` will be `NULL`.

**Comment by vladima** [comment]
thanks! Can you please tell when these changes might land in Node once this PR is merged in?

**Comment by bnoordhuis** [comment]
We'll do a libuv release once enough significant changes have piled up and upgrade the bundled copy in nodejs/node afterwards.  That in turn is released in the next Current (v7.x) release and a few weeks or months later in LTS releases (v4.x and v6.x.)

**Comment by vladima** [comment]
I see that CI build fails for linux and freebsd configurations with some error inside Hudson. Is there anything that I need to do on my side to make it pass?

**Comment by santigimeno** [comment]
The failures look unrelated.

**Comment by santigimeno** [comment]
Landed in https://github.com/libuv/libuv/commit/02dcde08386441d5a89dbcb602a1ad367a506cc0 with the commits squashed and the commit message formatted. Thanks!

**Pull request closed without merging**